### ISSUE TITLE
Stop using floats to avoid rounding issues

### ISSUE THREAD

**Issue by zorun** [spiral-project/ihatemoney#165]
Using floats for bill amount leads to all sorts of rounding issues, see for instance #126 and #138.  Basically, computations with floats are not exact, because it is not possible to represent exactly some decimal numbers in a binary system.  For instance with python 2.7.13 and 3.5.2:

```
In [1]: 0.04 + 0.07 - 0.11
Out[1]: 1.3877787807814457e-17
```

One solution would be to use a [Numeric](http://docs.sqlalchemy.org/en/latest/core/type_basics.html#sqlalchemy.types.Numeric) type for the database, which gets mapped to [decimal.Decimal](https://docs.python.org/2/library/decimal.html) objects in Python. However, sqlite only supports floats, so using a Numeric datatype would not change anything when using sqlite:

```
sqlalchemy/sql/sqltypes.py:596: SAWarning: Dialect sqlite+pysqlite does *not* support Decimal objects natively, and SQLAlchemy must convert from floating point - rounding errors and other issues may occur. Please consider storing Decimal numbers as strings or integers on this platform for lossless storage.
```

An alternative could be to switch to integers (as suggested by the warning), for instance storing a number of cents as integer.

However, this would bring another issue when splitting the amount of a bill.  For instance, splitting `1.42` in 5 equal parts gives either `0.28` or `0.29`.  In the first case, 2 cents would disappear, while in the second case, 3 cents would be created.  The current method of using floats avoids this issue.

In short: numbers are not easy :)

**Comment by JocelynDelalande** [comment]
@zorun thanks for taking time to explain the issue, I learnt things :-)

**Comment by almet** [comment]
Hi, thanks for taking the time to explain this so precisely!

> However, this would bring another issue when splitting the amount of a bill. For instance, splitting 1.42 in 5 equal parts gives either 0.28 or 0.29. In the first case, 2 cents would disappear, while in the second case, 3 cents would be created.

I would personally go in this direction: we don't really care about cents here, since they're cents (and that would make no real difference) and they are used only for display (as a way for users to roughly know the parts).

**Comment by almet** [comment]
cc @feth in case you want to have a look.

**Comment by FabriceSalvaire** [comment]
I am just reading this issue, **and I think float (double) are not an issue at all.** 

The issue is due to cent which are only represented by 2 decimal.

1.42 / 5 is exactly equal to 0.284 and thus 5 * 0.004 € is lost by the cent rounding.

For each transaction, we are always losing or winning some cents which should be equilibrated over the time.

However there are some rules to follow like apply VAT on the total and not on items in order to limit cent rounding side effect. **For ihatemoney,  cent rounding must be done at last.** Double precision division inaccuracy should be negligible for most bills.

**Comment by Glandos** [comment]
I am currently hitting this issue, since the total balance is `9.4449999999997` in the API call.
Maybe it could be stored twice, one as float for direct DB operations, and one as string for Python operations as Decimal.

**Comment by almet** [comment]
Hi everyone, if someone want to have a closer look and propose a fix for this, I will happily review it and provide feedback.

Cheers!

**Comment by almet** [comment]
We're now using an external lib for this. Closing.
